FEATURE: Add Fusion prototypes `Join`, `Loop`, `Map`, `Reduce` and `DataStructure`

mficzel opened the pull request:
The difference between ``RawArray``, ``Array``, ``RawCollection`` and ``Collection`` was often hard to get for new developers. To overcome this, the old confusing names are deprecated, and new prototypes are introduced which are easier to understand and emphasize the declarative nature of Fusion.

``Neos.Fusion:Join``

The prototype concatenates the fusion values of all fusion-properties and returns the result as a string. This prototype replaces ``Neos.Fusion:Array`` which is deprecated.

In addition to the behavior of the ``Array`` ``Join`` allows to define the ``@glue`` used for concatenating the parts.

``Neos.Fusion:Loop``

The ``Loop`` prototype iterates over the given items with the itemRenderer and returns the concatenated result as a string. The prototype replaces ``Neos.Fusion:Collection`` which is deprecated.

Other than in ``Collection`` the items are passed with the key ``items`` instead of ``collection``.

In addition to the behavior of the classic ``Collection`` ``Loop ``  allows defining the ``@glue`` used for concatenating the items. 

``Neos.Fusion:DataStructure``

The ``DataStructure`` prototype returns an associative array with all fusion keys evaluated. The prototype replaces ``Neos.Fusion:RawArray`` which is deprecated.

``Neos.Fusion:Map``

The ``Map`` iterates over the given ``items`` and returns the result as an array.  The prototype replaces ``Neos.Fusion:RawCollection`` which is deprecated.

Other than ``RawCollection`` the items are passed with the key ``items`` instead of ``collection`` and the keys of the given ``items`` are preserved.

``Neos.Fusion:Reduce``

The `Neos.Fusion:Reduce` prototype is added which reduces the given items to a single value by using ``itemRenderer`` with the following properties.

- `items` (array/Iterable, **required**) The array or iterable to iterate over
- `itemName`: (string, defaults to `item`) Context variable name for each item
- `itemKey`: (string, defaults to `itemKey`) Context variable name for each item key, when working with an array
- `carryName`: (string, defaults to `carry`) Context variable that contains the result of the last iteration
- `iterationName`: (string, defaults to `iterator`) A context variable with iteration information will be available under the given name: ``index`` (zero-based), ``cycle`` (1-based), ``isFirst``, ``isLast``
- `itemReducer`: (mixed, **required**) The reducer definition (simple value, expression or object) that will be applied for every item.
- `initialValue`: (mixed, defaults to `null`) The value that is passed to the first iteration or returned if the items are empty

daniellienert commented on Neos.Neos/Documentation/References/NeosFusionReference.rst:
```diff
@@ -364,6 +424,21 @@ Evaluate nested definitions as an array (opposed to *string* for :ref:`Neos_Fusi
 
 .. tip:: For simple cases an expression with an array literal ``${[1, 2, 3]}`` might be easier to read
 
+.. note:: The Neos.Fusion:RawArray object has been renamed to Neos.Fusion:Dictionary the old name is &lt;b&gt;(DEPRECATED)&lt;/b&gt;
```
Is `&lt;b&gt;` intended here?

mficzel replied:
I simply found that in another deprecation notice. Not much intention from my side other than that

daniellienert commented on Neos.Fusion/Classes/FusionObjects/CollectionImplementation.php:
```diff
@@ -18,6 +18,7 @@
  *
  * //tsPath collection *Collection
  * //tsPath itemRenderer the TS object which is triggered for each element in the node collection
+ * @deprecated since Neos 4.2 in favor of IterationImplementation
```
Would be good to name the version, where it will be removed (5.0 ?)

mficzel replied:
I am not sure. I think we should allow some time util we remove those prototypes since.

jonnitto replied:
I would wait until version 6 since it is a massive change. Especially for package developer: They would have to handle multiple versions. If we remove it later, the transition would be smoother

albe replied:
Doesn't a simple renaming code migration solve the major issue? AFAIS the usages are fully b/c otherwise apart from the namechange, no?

daniellienert commented on Neos.Neos/Documentation/References/NeosFusionReference.rst:
```diff
@@ -106,6 +117,8 @@ Render each item in ``collection`` using ``itemRenderer``.
 :iterationName: (string, defaults to ``iterator``) A context variable with iteration information will be available under the given name: ``index`` (zero-based), ``cycle`` (1-based), ``isFirst``, ``isLast``
 :itemRenderer: (string, **required**) The renderer definition (simple value, expression or object) will be called once for every collection element, and its results will be concatenated
 
+.. note:: The Neos.Fusion:Collection object is &lt;b&gt;(DEPRECATED)&lt;/b&gt; use Neos.Fusion:Iteration instead.
```
Is &lt;b&gt; intended here?

daniellienert commented:
Hey @mficzel, 
Thats a good idea! Thanks. It so sad that array is taken by the wrong thing. `Dictionary` is an unknown term both in PHP and JS - but I wasn't able to find something better.

Have you thought about writing migrations?

mficzel commented:
No migrations yet as I plan to introduce this as a feature in 4.2. Migrations should be in 5.0 when we remove the old names and also adjust the fusion code in the Neos.Neos package. For now this is just an extra option

mficzel commented:
Maybe it is wise if we still keep the names until 6.0 but add the migrations at still in version 5 and stop using the old names

dimaip commented:
Wow, great initiative!
The naming sounds a bit weird to me, but it may be hard to come up with anything much better.
`Concat` is a good name, nothing against it.
`Iteration` is not just an abstract iteration, but actually mapping + concat, so could be great if they were named somehow similarly.
`Dictionary` is an unfamiliar term, but not sure if there are better alternatives either, since both collection and array are already taken.

mficzel commented:
@dimaip `Iteration` makes a lot of sense in AFX but i am totally open to better names. I also plan to add a `Reduction` prototype soonish.

dimaip commented:
Yep, as said, I don't have anything better to offer, so I just shared my concerns. Maybe somebody would have ideas, if not we could leave it as you suggest.

mficzel commented:
@dimaip this is based on the discussion here https://discuss.neos.io/t/rfc-cleanup-fusion-prototype-names/3646/11

dimaip commented:
Ah, already forgot about that discussion. But there was an idea to rename `RawArray` into `Map`, why did you give up on it?

albe commented on Neos.Fusion/Classes/FusionObjects/DictionaryImplementation.php:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Evaluate sub objects to an array (instead of a string as ConcatImplementation does)
+ */
+class DictionaryImplementation extends ArrayImplementation
```
Shouldn't `DictionaryImplementation` extend `ConcatenationImplementation` directly, instead of the deprecated `ArrayImplementation`? But then again, a `Dictionary` really isn't a `Concatenation` in any way, so this hints at some improper naming or inheritance. So what's the common ground? The sorting. Actually, `Dictionary` doesn't do much else. What about `Dictionary` becoming a `Sorting` and `Concatenation` extending that?

mficzel replied:
@albe the names try to describe the expected result instead of the operation or process. Imho this is what reflects the declarative nature of fusion best.

albe reviewed (commented):
One thing that kind of bugs me:

`Concatenation`, `Mapping`, `Iteration`, `Reduction` are all names for a process. `Dictionary` is a name for a structure. This some how feels out of line.

While looking at the code, I noticed that `Dictionary` is actually about sorting the elements. So what about calling it `Sorting`? (see also other comment as to why that might further sense)

mficzel commented:
@albe Thank you for the hint regarding inheritence, i adjusted the code accordingly. The reason why it looked like this in the first place was that RawArray also was based on array. A good timt to fix this.

Regarding the naming i tried to name the objects after the expected result and not the process or the operation. That is why i ended up with `Dictionary`. 

In general i think we should emphasize the declarative nature of fusion way more. Fusion Objects encapsulate the imperative logic and allow the Integrators to declare the requirements.

albe commented:
Thanks. That looks better already. So what actually speaks against `Sorting`? The result is a sorted associative array. Does it not fit? I'm not really deep into Fusion, so maybe I miss something.

mficzel commented:
@albe `Sorting` is the process. If we want to describe the result `SortedAssociativeArray`would be a better name but is quite unhandy. A Dictionary has keys and values in a specific order, to me this sounds quite good.

`Array`(PHP) or `Object`(JS) would be the really best names ... however those are not available.

albe commented:
Sorry if I'm a bit thick, but I still don't understand why all the other Fusion Prototypes got names that describe the process. `Iteration` instead of `Iterable`, `Mapping` instead of `Collection`, etc. - in those cases the names also do not describe the result, otherwise `Concatenation` should be named `String`, no?

If it stays as a structural name, generally I'd also prefer `Map` over `Dictionary`, because it's more known from other languages to me, coming more from a C/Java background (https://en.wikipedia.org/wiki/Associative_array#Language_support). BUT: `Map` can also be the name of the higher order function, as in map/reduce, but the supposed counterpart is named `Reduction` - this would raise questions. Also, it would create misunderstanding together with `Mapping`, therefore `Dictionary` is probably the best choice. However, I still feel this misses the significant point, that the result is an ordered map/dictionary, which is normally not guaranteed by a dictionary/map.

mficzel commented:
@albe i totally understand the need to get this right and appreciate your effort. So please be thick. This pr tries to overcome a bad naming situation and i hope to not make things worse.

It could work to use `Map`, `Iterate`, `Dictionary`, `Concat` and `Reduce` however in this case the term Map is a bit confusing. Mapping is clear related to the transformation of items. 

Regarding iteration i suggest to look at it from the presentation side where compnents are defined in afx and take an array of dictionaries as input.

```
afx`
<ul>
<Neos.Fusion:Iteration items={props.items}>
   <li>{item.title}</li>
</Neos.Fusion:Iteration>
</ul>`
```
That is why i decided to use `Iteration` but i am totally open for better suggestions.

In general we should make clear that we are not exposing php functionality to a template language fluid did in the end. We create fusion to allow to describe the result, that is why i would not like names like `Iterable`.

danielkestler commented:
Sorry, but I have to say: the old names are more logical to me. While I get Dictionary (from an Objective-C background), the others seem weird.

But good effort anyways, because the names indeed need some work as the old names are also not self explanatory.

danielkestler commented:
Leaving backwards compatibility aside, just some ideas:

`Neos.Fusion:Array` and `Neos.Fusion:ArrayString`
`Neos.Fusion:Iteration` and `Neos.Fusion:IterationString`

... or `...ToString` / `...AsString` / `...StringRepresentation`?

danielkestler commented:
We could also leave the string variants behind and promote to use a Stringify-processor.

kitsunet commented:
> We could also leave the string variants behind and promote to use a Stringify-processor.

I like that idea. Alternatively Concatenation could also be something like a `Join` or `JoinStrings`.

Although `Concatenation` works fine for me.

`Neos.Fusion:Dictionary` could maybe be a `DataStructure`? That is pretty straight forward and not as theoretical or even a `DataObject`. Or `KeyValueMap`, just to throw in some more options. I don't dislike Dictionary but I feel it's rather abstract and theoretical.

Technically `Iteration` is a specialised `Reduction` right? That feels slightly off. Maybe very expressive `RenderAndJoin`. I think Iteration is the one I am most on the fence about but also the one I really have no great idea about... I will let it sink a bit more.

Mapping works pretty nicely I think. Reduction also works fine IMHO. Not sure about performance and caching implications, but seems like a nice addition.

kitsunet commented:
Additional note. We must make sure we don't make it all too technical. Fusion is meant to be an integrators tool after all. Yes you would usually expect some JS knowledge and with that an understanding for the used terminology BUT I would try to avoid making it too technical if possible. :)

mficzel commented:
@kitsunet `DataStructure` sounds nice for me aswell. If others like it aswell i will adjust the code.

Regarding `Concatenation` vs `Join` i think that the latter should offer a `glue`-option wich will not work with the current way we use Fusion `Array`. If we have an idea how to configure the glue (maybe via @context.glue="") we should make it configurable and maybe use `Join`. However `Join` does not describe the result but the operation so we should look at it from the integration standpoint and describe the result.

For the `Iteration` it is the same. The most important question for me is what looks intuitive in afx when we do things like this:
```
<ul><Neos.Fusion:Iteration items={props.item}>
  <li @path="itemRenderer">{item.title}</li>
</Neos.Fusion:Iteration></ul>
```

mficzel commented:
An option that just came to my mind as possible alternative to Iteration is `Loop` that would look like this.
```
<ul><Neos.Fusion:Loop items={props.item}>
  <li @path="itemRenderer">{item.title}</li>
</Neos.Fusion:Loop></ul>
```
I would love to support `content` as fallback to `itemRenderer` to avoid the @chidren or @path annotation.

kitsunet commented:
@mficzel why wouldn't the Join thing work with how we use array? We could just add the glue option with a default that is the same that we do internally at the moment.

mficzel commented:
@kitsunet we expect all keys of an arrayFusionObject to be rendered so there is no possible key we can use to define the glue for the join. It also feels wrong to have an api where all keys but glue do not matter.

I would consider a configureable glue important for a thing called `Join` but not for a `Concatenation`. However if we have a good idea how to configure the glue calling the thing `Join` is a good idea.

bwaidelich commented:
I agree that we should look at this from the integrators point of view to find out what "feels" right and obviously don't use technical terms for something they don't represent.

To wrap my head around this I had to come up with a (frankly somewhat wacky) example that uses the prototypes in question:
A person component that renders the full name and a list of friends

### Today

<pre><code>prototype(My.Package:Person) &#x3C; prototype(Neos.Fusion:Component) {
    givenName = &#x27;&#x27;
    familyName = &#x27;&#x27;
    friends = <b>Neos.Fusion:RawArray</b>

    renderer = afx&#x60;
        &#x3C;section&#x3E;
            &#x3C;h1&#x3E;&#x3C;My.Package:Person.Name givenName={props.givenName} familyName={props.familyName} /&#x3E;&#x3C;/h1&#x3E;
            &#x3C;ul @if.hasFriends=&#x22;{!Array.empty(props.friends)}&#x22;&#x3E;
                &#x3C;<b>Neos.Fusion:Collection</b> collection={props.friends} itemName=&#x22;friend&#x22; @children=&#x22;itemRenderer&#x22;&#x3E;
                    &#x3C;li&#x3E;&#x3C;My.Package:Person.Name givenName={friend.givenName} familyName={friend.familyName} /&#x3E;&#x3C;/li&#x3E;
                &#x3C;/Neos.Fusion:Collection&#x3E;
            &#x3C;/ul&#x3E;
        &#x3C;/section&#x3E;
    &#x60;
}

prototype(My.Package:Person.Name) &#x3C; prototype(Neos.Fusion:Component) {
    givenName = &#x27;&#x27;
    familyName = &#x27;&#x27;

    renderer = <b>Neos.Fusion:Array</b> {
        givenName = ${props.givenName}
        space = &#x27; &#x27;
        space.@if.hasFamilyName = ${props.familyName}
        familyName = ${props.familyName}
    }
}</code></pre>
#### Usage:

<pre><code>root = My.Package:Person {
    givenName = ${q(personNode).property(&#x27;givenName&#x27;)}
    familyName = ${q(personNode).property(&#x27;familyName&#x27;)}
    friends = <b>Neos.Fusion:RawCollection</b> {
        collection = ${q(personNode).property(&#x27;friends&#x27;)}
        itemName = &#x27;friendNode&#x27;
        itemRenderer = <b>Neos.Fusion:RawArray</b> {
            givenName = ${q(friendNode).property(&#x27;givenName&#x27;)}
            familyName = ${q(friendNode).property(&#x27;givenName&#x27;)}
        }
    }
}</pre></code>

### Suggestion

<pre><code>prototype(My.Package:Person) &#x3C; prototype(Neos.Fusion:Component) {
    givenName = &#x27;&#x27;
    familyName = &#x27;&#x27;
    friends = ${[]}

    renderer = afx&#x60;
        &#x3C;section&#x3E;
            &#x3C;h1&#x3E;&#x3C;My.Package:Person.Name givenName={props.givenName} familyName={props.familyName} /&#x3E;&#x3C;/h1&#x3E;
            &#x3C;ul @if.hasFriends=&#x22;{!Array.empty(props.friends)}&#x22;&#x3E;
                &#x3C;<b>Neos.Fusion:Loop</b> items={props.friends} itemName=&#x22;friend&#x22;&#x3E;
                    &#x3C;li&#x3E;&#x3C;My.Package:Person.Name givenName={friend.givenName} familyName={friend.familyName} /&#x3E;&#x3C;/li&#x3E;
                &#x3C;/Neos.Fusion:Loop&#x3E;
            &#x3C;/ul&#x3E;
        &#x3C;/section&#x3E;
    &#x60;
}

prototype(My.Package:Person.Name) &#x3C; prototype(Neos.Fusion:Component) {
    givenName = &#x27;&#x27;
    familyName = &#x27;&#x27;

    renderer = <b>Neos.Fusion:Concat</b> {
        givenName = ${props.givenName}
        space = &#x27; &#x27;
        space.@if.hasFamilyName = ${props.familyName}
        familyName = ${props.familyName}
    }
}</pre></code>

#### Usage:

<pre><code>root = My.Package:Person {
    givenName = ${q(personNode).property(&#x27;givenName&#x27;)}
    familyName = ${q(personNode).property(&#x27;familyName&#x27;)}
    friends = <b>Neos.Fusion:Mapping</b> {
        collection = ${q(personNode).property(&#x27;friends&#x27;)}
        itemName = &#x27;friendNode&#x27;
        itemRenderer = <b>Neos.Fusion:Dictionary</b> {
            givenName = ${q(friendNode).property(&#x27;givenName&#x27;)}
            familyName = ${q(friendNode).property(&#x27;givenName&#x27;)}
        }
    }
}</pre></code>

### My feelings after this experiment

* `Loop` works very well for me (but `Iteration` would be fine, too)
* I like `Concat` as replacement for `Array` (but `Join` would work, too. And some "glue" would be nice for either and would make the `space.@if` condition above obsolete. But IMO that could be added as a separate feature)
* I prefer `Dictionary` over `Map` for the reason mentioned [here](https://stackoverflow.com/questions/2884068/what-is-the-difference-between-a-map-and-a-dictionary)


One more point regarding the latter:
I replaced the `friends = Neos.Fusion:RawArray` with `friends = ${[]}` above because `Dictionary` would be very confusing there.
That made me think: Maybe we can add a similar syntactic sugar for associative arrays to Fusion:

<pre><code> itemRenderer = Neos.Fusion:Dictionary {
  givenName = //...
  familyName = //...
}</pre></code>

=> 

<pre><code> itemRenderer = {
  givenName = //...
  familyName = //...
}</pre></code>

but that's out of scope of this change

mficzel commented:
In a hangout we agreed on the following names
```
Neos.Fusion:Array => Neos.Fusion:Join (with @glue)
Neos.Fusion:Collection => Neos.Fusion:Loop (with @glue)
Neos.Fusion:RawArray => Neos.Fusion:DataStructure (collection > items)
Neos.Fusion:RawCollection => Neos.Fusion:Map (collection > items)
---  new --- => Neos.Fusion:Reduce 
``` 
I will adjust the pr sonnish to that

bwaidelich commented:
^ as discussed with @mficzel @kitsunet and @danielkestler 

PS: @mficzel the `(collection > items)` at DataStructure belongs to `Loop` :)

With the suggested names, my example above would look like this:

<pre><code>prototype(My.Package:Person) &#x3C; prototype(Neos.Fusion:Component) {
    givenName = &#x27;&#x27;
    familyName = &#x27;&#x27;
    friends = ${[]}

    renderer = afx&#x60;
        &#x3C;section&#x3E;
            &#x3C;h1&#x3E;&#x3C;My.Package:Person.Name givenName={props.givenName} familyName={props.familyName} /&#x3E;&#x3C;/h1&#x3E;
            &#x3C;ul @if.hasFriends=&#x22;{!Array.empty(props.friends)}&#x22;&#x3E;
                &#x3C;<b>Neos.Fusion:Loop</b> items={props.friends} itemName=&#x22;friend&#x22;&#x3E;
                    &#x3C;li&#x3E;&#x3C;My.Package:Person.Name givenName={friend.givenName} familyName={friend.familyName} /&#x3E;&#x3C;/li&#x3E;
                &#x3C;/Neos.Fusion:Loop&#x3E;
            &#x3C;/ul&#x3E;
        &#x3C;/section&#x3E;
    &#x60;
}

prototype(My.Package:Person.Name) &#x3C; prototype(Neos.Fusion:Component) {
    givenName = &#x27;&#x27;
    familyName = &#x27;&#x27;

    renderer = <b>Neos.Fusion:Join</b> {
        @glue = ' '
        givenName = ${props.givenName}
        familyName = ${props.familyName}
    }
}</pre></code>

#### Usage:

<pre><code>root = My.Package:Person {
    givenName = ${q(personNode).property(&#x27;givenName&#x27;)}
    familyName = ${q(personNode).property(&#x27;familyName&#x27;)}
    friends = <b>Neos.Fusion:Map</b> {
        items = ${q(personNode).property(&#x27;friends&#x27;)}
        itemName = &#x27;friendNode&#x27;
        itemRenderer = <b>Neos.Fusion:DataStructure</b> {
            givenName = ${q(friendNode).property(&#x27;givenName&#x27;)}
            familyName = ${q(friendNode).property(&#x27;givenName&#x27;)}
        }
    }
}</pre></code>

I like

bwaidelich commented on Neos.Fusion/Classes/FusionObjects/JoinImplementation.php:
```diff
@@ -0,0 +1,46 @@
+     */
+    public function getGlue()
+    {
+        return $this->fusionValue('__meta/glue') ?: '';
```
Shouldn't that use the null coalescing operator instead?
```suggestion
        return $this->fusionValue('__meta/glue') ?? '';
```

bwaidelich replied:
Nah, nevermind, should not make a difference in this case. (I just wanted to use the new "suggest change" feature :)

mficzel commented:
This pr is imho ready now. Please review and let us get this finally merged.

jonnitto reviewed (approved):
Looks very nice! 👍

> I approve it only from a conceptual point of view, as I am not a PHP developer

mficzel committed:
FEATURE: Add Fusion prototypes `Join`, `Loop`, `Map`, `Reduce` and `DataStructure`

The difference between ``RawArray``, ``Array``, ``RawCollection`` and ``Collection`` was often hard to get for new developers. To overcome this the old confusing names are deprecated and new prototypes are introduced wich are easier to get and emphasize the declarative nature of Fusion.

``Neos.Fusion:Join``

The prototype concatenates the fusion values of all fusion-properties and returns the result as a string. This prototype replaces ``Neos.Fusion:Array`` wich is deprecated.

In addition to the behavior of the ``Array`` ``Join`` allows to define the ``@glue`` used for concatenating the parts.

``Neos.Fusion:Loop``

The ``Loop`` prototype iterates over the given items with the itemRenderer and returns the concatenated result as a string. The prototype replaces ``Neos.Fusion:Collection`` wich is deprecated.

Other than in ``Collection`` the items are passed with the key ``items`` instead of ``collection``.

In addition to the behavior of the classic ``Collection`` ``Loop ``  allows to define the ``@glue`` used for concatenating the items.

``Neos.Fusion:DataStructure``

The ``DataStructure`` prototype returns an associative array with all fusion keys evaluated. The prototype replaces ``Neos.Fusion:RawArray`` wich is deprecated.

``Neos.Fusion:Map``

The ``Map`` iterates over the given ``items`` and returns the result as array.  The prototype replaces ``Neos.Fusion:RawCollection`` wich is deprecated.

Other than ``RawCollection`` the items are passed with the key ``items`` instead of ``collection`` and the keys of the given ``items`` are preserved.

``Neos.Fusion:Reduce``

The `Neos.Fusion:Reduce` prototype is added wich reduces the given items to a single value by using ``itemRenderer`` with the following properties.

- `items` (array/Iterable, **required**) The array or iterable to iterate over
- `itemName`: (string, defaults to `item`) Context variable name for each item
- `itemKey`: (string, defaults to `itemKey`) Context variable name for each item key, when working with array
- `carryName`: (string, defaults to `carry`) Context variable that contains the result of the last iteration
- `iterationName`: (string, defaults to `iterator`) A context variable with iteration information will be available under the given name: ``index`` (zero-based), ``cycle`` (1-based), ``isFirst``, ``isLast``
- `itemReducer`: (mixed, **required**) The reducer definition (simple value, expression or object) that will be applied for every item.
- `initialValue`: (mixed, defaults to `null`) The value that is passed to the first iteration or returned if the items are empty

jonnitto commented:
@kitsunet @mficzel Would it make sense to update the used prototypes in the development collection?

mficzel commented:
@jonnito I plan to remove our usages of the old fusion-names in Neos 5. Feels more appropriate to do this in a major and now just introduce the new prototypes as an option. However if anyone wants do this faster i will not object.

manuelmeister commented:
When do we delete these deprecated prototypes? For newcomers it's quite difficult to see so many (from other languages familiar: Array, RawArray, Collection) named prototypes just to realize they are all deprecated.

mficzel commented:
@manuelmeister would rather make the deprecation clearly visible. Do you have a suggestion how that should be done to be clear for newcomers.

manuelmeister commented:
@mficzel made a suggestion here: https://discuss.neos.io/t/make-the-fusion-api-references-more-newcomer-friendly/6132
